pkg/execution/svccatadmission/internaldns.go

```diff
@@ -0,0 +1,141 @@
+
+	"github.com/atlassian/voyager/pkg/k8s"
+	"github.com/atlassian/voyager/pkg/microsserver"
+	"github.com/atlassian/voyager/pkg/orchestration/wiring/internaldns/api"
+	"github.com/atlassian/voyager/pkg/util/logz"
+	sc_v1b1 "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	"github.com/pkg/errors"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	ext_v1b1 "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InternalDNSAdmitFunc checks existing DNS alias ownership via micros server API, for both InternalDNS Services and Ingress Resources
+func InternalDNSAdmitFunc(ctx context.Context, microsServerClient microsServerClient, scClient serviceCentralClient, admissionReview admissionv1beta1.AdmissionReview) (*admissionv1beta1.AdmissionResponse, error) {
+	logger := logz.RetrieveLoggerFromContext(ctx).Sugar()
+	admissionRequest := admissionReview.Request
+
+	// populating a chanel of domains to check owner
+	var domainsToCheck chan string
+	switch admissionRequest.Resource {
+	case k8s.IngressGVR:
+		ingress := ext_v1b1.Ingress{}
+		if err := json.Unmarshal(admissionRequest.Object.Raw, &ingress); err != nil {
+			return nil, errors.Errorf("error parsing Ingress resource")
+		}
+		if len(ingress.Spec.Rules) == 0 {
+			return nil, errors.Errorf("cannot process Ingress with empty rules list")
+		}
+		domainsToCheck = make(chan string, len(ingress.Spec.Rules))
+		for _, ingressRule := range ingress.Spec.Rules {
+			domainsToCheck <- ingressRule.Host
+		}
+	case k8s.ServiceInstanceGVR:
+		serviceInstance := sc_v1b1.ServiceInstance{}
+		if err := json.Unmarshal(admissionRequest.Object.Raw, &serviceInstance); err != nil {
+			return nil, errors.Wrap(err, "error parsing ServiceInstance")
+		}
+		var internalDNSSpec apiinternaldns.Spec
+		if err := json.Unmarshal(serviceInstance.Spec.Parameters.Raw, &internalDNSSpec); err != nil {
+			return nil, errors.Wrap(err, "error parsing InternalDNS spec")
+		}
+		if len(internalDNSSpec.Aliases) == 0 {
+			return nil, errors.Errorf("cannot process InternalDNS with empty aliases list")
+		}
+		domainsToCheck = make(chan string, len(internalDNSSpec.Aliases))
+		for _, alias := range internalDNSSpec.Aliases {
+			domainsToCheck <- alias.Name
+		}
+	default:
+		return nil, errors.Errorf("unsupported resource, got %v", admissionRequest.Resource)
+	}
+	close(domainsToCheck)
+
+	// fetching service data from service central
+	serviceName := getServiceNameFromNamespace(admissionRequest.Namespace)
+	serviceCentralData, err := getServiceData(ctx, scClient, serviceName)
+	if err != nil {
+		return nil, errors.Errorf("error fetching service central data for serviceName %q", serviceName)
+	}
+
+	// types to store parallel domain owner fetches
+	type domainOwnership struct {
+		requestedDomain         string
+		existingDomainAliasInfo *microsserver.AliasInfo
+		err                     error
+	}
+	domainsOwnership := make([]domainOwnership, cap(domainsToCheck))
```

Comment:
```suggestion
	domainsOwnership := make([]domainOwnership, 0, cap(domainsToCheck))
```
Otherwise you have a non-empty slice.

Reply:
done, pushed